src/services/pcn-bridge/src/Bridge_dp.c

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright 2019 The Polycube Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bcc/helpers.h>
+#include <bcc/proto.h>
+
+#include <uapi/linux/bpf.h>
+#include <uapi/linux/filter.h>
+#include <uapi/linux/if_ether.h>
+#include <uapi/linux/if_packet.h>
+#include <uapi/linux/if_vlan.h>
+
+#include <uapi/linux/in.h>
+#include <uapi/linux/ip.h>
+#include <uapi/linux/pkt_cls.h>
+
+enum { BPDU_REASON = 1, BROADCAST_REASON };
+
+enum { PORT_MODE_ACCESS = 0, PORT_MODE_TRUNK };
+
+enum entry_type { STATIC = 0, DYNAMIC };
+
+#define VLAN_WILDCARD 0x00
+
+#if STP_ENABLED
+enum stp_state {
+  STP_DISABLED = 1 << 0,   /* 8.4.5: See note above. */
+  STP_LISTENING = 1 << 1,  /* 8.4.2: Not learning or relaying frames. */
+  STP_LEARNING = 1 << 2,   /* 8.4.3: Learning but not relaying frames. */
+  STP_FORWARDING = 1 << 3, /* 8.4.4: Learning and relaying frames. */
+  STP_BLOCKING = 1 << 4    /* 8.4.1: Initial boot state. */
+};
+
+/* Returns true if 'state' is one in which packets received on a port should
+ * be forwarded, false otherwise.
+ *
+ * Returns true if 'state' is STP_DISABLED, since presumably in that case the
+ * port should still work, just not have STP applied to it. */
+static __always_inline u32 stp_forward_in_state(u32 state) {
+  return (state & (STP_DISABLED | STP_FORWARDING)) != 0;
+}
+
+/* Returns true if 'state' is one in which MAC learning should be done on
+ * packets received on a port, false otherwise.
+ *
+ * Returns true if 'state' is STP_DISABLED, since presumably in that case the
+ * port should still work, just not have STP applied to it. */
+static __always_inline u32 stp_learn_in_state(u32 state) {
+  return (state & (STP_DISABLED | STP_LEARNING | STP_FORWARDING)) != 0;
+}
+#endif
+
+/*
+ * Allowed Vlans table is used to set and verify vlans associated to a port
+ * in access or trunk mode. The usage is provided in the following example.
+ */
+struct port_vlan_key {
+  u32 port;
+  u16 vlan;
+} __attribute__((packed));
+
+struct port_vlan_value {
+  u16 vlan;
+  u16 stp_state;
+} __attribute__((packed));
+
+BPF_TABLE("hash", struct port_vlan_key, struct port_vlan_value, port_vlan,
+          1024);
+/*
+ * port_vlan usage example:
```

Comment:
Honestly, the above comments in the code are not clear to me; it is difficult to understand what this table is about.

Reply:
I didn't write these comments and i had the same difficulties to understand the purpose of the table.
In short, in this table we can check if a vlan is allowed for a certain port (if, for a given key, the value is present so the vlan is allowed). It's also useful to know the state of a port when stp is enabled.
Maybe i should add this explanation as a comment in the code

Reply:
Yes, please. Thanks!